File: J9Ch/src/S_3_Data_Structure_II/Required_11/_126_Max_Tree_Hard.py

"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        this.val = val
        this.left, this.right = None, None
"""
class Stack:
    def __init__(self):
        self.items = []

    def __getitem__(self, index):
        return self.items[index]

    def isEmpty(self):
        return len(self.items)==0

    def push(self, item):
        self.items.append(item)

    def pop(self):
        return self.items.pop()

    def peek(self):
        if not self.isEmpty():
            return self.items[len(self.items)-1]

    def size(self):
        return len(self.items)
class Solution:
    # @param A: Given an integer array with no duplicates.
    # @return: The root of max tree.
    def maxTree(self, A):
        # write your code here
        stk = Stack()
        for ele in A:
            node = TreeNode(ele)
            while not stk.isEmpty() and ele > stk.peek().val:
                node.left = stk.pop()
            if not stk.isEmpty():
                stk.peek().right = node
            stk.push(node)
        return stk[0]

File: JS9Ch/src/S_3_Data_Structure_II/Optional_13/python_cpp/_126_Max_Tree_Hard.cpp


/**
 * Definition of TreeNode:
 * class TreeNode {
 * public:
 *     int val;
 *     TreeNode *left, *right;
 *     TreeNode(int val) {
 *         this->val = val;
 *         this->left = this->right = NULL;
 *     }
 * }
 */
class Solution {
public:
    /**
     * @param A: Given an integer array with no duplicates.
     * @return: The root of max tree.
     */
    vector<TreeNode *> stk;
    TreeNode* maxTree(vector<int> A) {
        // write your code here
        int len = A.size();
        for (int i = 0; i < len; ++i) {
            TreeNode *tmp = new TreeNode(A[i]);
            while (stk.size() > 0 && A[i] > stk[stk.size()-1]->val) {
                tmp->left = stk[stk.size()-1];
                stk.pop_back();
            }
            if (stk.size() > 0)
                stk[stk.size()-1]->right = tmp;
            stk.push_back(tmp);
        }
        return stk[0];
    }
};